target/riscv/cpu.c

```diff
@@ -364,43 +364,139 @@ void restore_state_to_opc(CPURISCVState *env, TranslationBlock *tb,
+    }
+    if (env->rvfi_dii_trace.available_fields & RVFI_MEM_DATA) {
+        g_byte_array_append(buf, (const guint8 *)"mem-data", 8);
+        g_byte_array_append(buf, (const guint8 *)&env->rvfi_dii_trace.MEM,
+                            sizeof(env->rvfi_dii_trace.MEM));
+    }
+    // Now that we know the total size, we can update the trace header:
+    ((struct rvfi_dii_trace_v2 *)buf->data)->trace_size = buf->len;
+    if (rvfi_debug_output) {
+        fprintf(stderr,
+            "Sending %u bytes: %jd PCWD: 0x%08jx, RD: %02d, RWD: 0x%08jx, MA: "
+            "0x%08jx, MWD: 0x%08jx, MWM: 0x%08x, I: 0x%016jx H:%u T:%u\n",
+            buf->len, (uintmax_t)env->rvfi_dii_trace.INST.rvfi_order,
+            (uintmax_t)env->rvfi_dii_trace.PC.rvfi_pc_wdata,
+            env->rvfi_dii_trace.INTEGER.rvfi_rd_addr,
+            (uintmax_t)env->rvfi_dii_trace.INTEGER.rvfi_rd_wdata,
+            (uintmax_t)env->rvfi_dii_trace.MEM.rvfi_mem_addr,
+            (uintmax_t)env->rvfi_dii_trace.MEM.rvfi_mem_wdata[0],
+            env->rvfi_dii_trace.MEM.rvfi_mem_wmask,
+            (uintmax_t)env->rvfi_dii_trace.INST.rvfi_insn,
+            (unsigned)env->rvfi_dii_trace.INST.rvfi_halt,
+            (unsigned)env->rvfi_dii_trace.INST.rvfi_trap);
+    }
+    send_rvfi_dii_packet(buf->data, buf->len);
+    g_byte_array_free(buf, true);
+}
+
+static void rvfi_dii_send_trace(CPURISCVState *env, unsigned version)
+{
+    if (version == 1) {
+        rvfi_dii_send_v1_trace(env);
+    } else if (version == 2) {
+        rvfi_dii_send_v2_trace(env);
+    } else {
+        error_report("Invalid trace version %d", version);
+        exit(EXIT_FAILURE);
+    }
+}
+
+void rvfi_dii_communicate(CPUState* cs, CPURISCVState* env, bool was_trap) {
+    // needs to be global since this function is called for each instruction
+    // that is executed.
     static bool rvfi_dii_started = false;
+    static unsigned rvfi_dii_version = 1;
     // Single-step completed -> update PC in the trace buffer
-    env->rvfi_dii_trace.rvfi_dii_pc_wdata = GET_SPECIAL_REG_ARCH(env, pc, PCC);
-    env->rvfi_dii_trace.rvfi_dii_order++;
+    env->rvfi_dii_trace.PC.rvfi_pc_wdata = GET_SPECIAL_REG_ARCH(env, pc, PCC);
+    env->rvfi_dii_trace.INST.rvfi_order++;
 
     // TestRIG expects a zero $pc after a trap:
-    if (env->rvfi_dii_trace.rvfi_dii_trap && rvfi_debug_output) {
+    if (env->rvfi_dii_trace.INST.rvfi_trap && rvfi_debug_output) {
         info_report("Got trap at " TARGET_FMT_lx, PC_ADDR(env));
     }
     env->rvfi_dii_have_injected_insn = false;
     while (true) {
         assert(cs->singlestep_enabled);
         rvfi_dii_command_t cmd_buf;
         _Static_assert(sizeof(cmd_buf) == 8, "Expected 8 bytes of data");
+        // Print the instruction now and skip the next commit() call that
+        // happens when we return to the translator loop.
+        qemu_log_instr_commit(env);
+        qemu_log_instr_drop(env); // Avoid an invalid instruction log
```

Comment:
@qwattash is this expected to work? The output looks sane to me and the next call to commit in the translator should clear the force_drop flag.

Reply:
I'm not familiar with the single stepping and RVFI_DII, how does this interact with the normal translator loop, are we injecting multiple instructions before we return to the translator loop?
I think this solution should work, but I think it will waste 1 slot in the ring buffer every time, if tracing in buffered mode.

Reply:
We inject one instruction, after that executes (but before the commit helper is called) it raises a singlestep trap.
That then calls this function. I added this here because I want the instruction changes to be logged before the rvfi packet data is sent so that I can compare what is being sent to what the log says.

Reply:
I see, so after the rvfi code returns we will go back to the loop and the normal `commit()` will be called. Using the force_drop flag here is fine. After checking, I don't think this will cause disruption to the ring buffer as the force_drop flag will cause the next commit to clear the buffer entry.